src/compat/posix/fs/dirent/dirent_oldfs.c

```diff
@@ -61,39 +62,67 @@ DIR *opendir(const char *path) {
 
 	d->path = node_path;
 	d->current.d_ino = 0;
+	d->dir_contex.fs_ctx = 0;
+
+	slist_init(&d->inodes_list);
 
 	return d;
 }
 
 int closedir(DIR *dir) {
+	struct inode *node;
+
 	if (NULL == dir) {
 		SET_ERRNO(EBADF);
 		return -1;
 	}
 
+	slist_foreach(node, &dir->inodes_list, dirent_link) {
+		inode_del(node);
+	}
 	objfree(&dir_pool, dir);
 
 	return 0;
 }
 
 struct dirent *readdir(DIR *dir) {
 	struct path child;
+	int res;
+	struct inode *node;
 
 	SET_ERRNO(0);
 
 	if (NULL == dir) {
 		SET_ERRNO(EBADF);
 		return NULL;
 	}
-
-	if ( 0 != vfs_get_child_next(&dir->path,
-			(struct inode *) (uintptr_t) dir->current.d_ino, &child)) {
-		return NULL;
+	res = vfs_get_child_next(&dir->path,(struct inode *) (uintptr_t) dir->current.d_ino, &child);
+	if (0 != res) {
+		if (dir->path.node->i_ops && dir->path.node->i_ops->iterate) {
+			node = inode_new(NULL);
+			if (NULL == node) {
+				SET_ERRNO(ENOMEM);
+				return NULL;
+			}
+			res = dir->path.node->i_ops->iterate(node, dir->path.node, &dir->dir_contex);
+			if (res == -1){
+				node_free(node);
+				return NULL;
+			}
```

Comment:
Should'nt we also do SET_ERRNO here like above?

Reply:
it's not a standard, and I don't know what code must be set, I return NULL, and it shows that function failed